### ISSUE TITLE
IllegalAccessError from lazy use of TransientLocalMemory

### ISSUE THREAD

**Issue by WilliamParker** [cerner/clara-rules#303]
I'd expect the following test to pass:

````
(def-rules-test test-negation-complex-join-with-numerous-non-matching-facts-inserted-after-descendant-negation
  {:rules []
   :queries [query1 [[]
                     [[Hot (= ?t temperature)]
                      [:not [Cold (join-filter-equals ?t temperature)]]]]]
   :sessions [empty-session [query1] {}]}
  (let [lots-of-hot (doall (for [_ (range 100)]
                             (->Hot 20)))]
    (is (= (repeat 100 {:?t 20})
           (-> empty-session
               (insert (->Cold 10))
               fire-rules
               (insert-all lots-of-hot)
               fire-rules
               (query query1))))))
````

Instead, on the JVM it throws:

````
java.lang.IllegalAccessError: Transient used after persistent!
 at clojure.lang.PersistentArrayMap$TransientArrayMap.ensureEditable (PersistentArrayMap.java:457)
    clojure.lang.ATransientMap.valAt (ATransientMap.java:78)
    clojure.lang.RT.get (RT.java:762)
    clara.rules.memory.TransientLocalMemory.get_elements (memory.cljc:441)
    clara.rules.engine.NegationWithJoinFilterNode$iter__16861__16865$fn__16866$fn__16867.invoke (engine.cljc:829)
    clara.rules.engine.NegationWithJoinFilterNode$iter__16861__16865$fn__16866.invoke (engine.cljc:826)
    clojure.lang.LazySeq.sval (LazySeq.java:40)
    clojure.lang.LazySeq.seq (LazySeq.java:49)
    clojure.lang.RT.seq (RT.java:521)
    clojure.core$seq__4357.invokeStatic (core.clj:137)
````

The full stack trace is in [this gist](https://gist.github.com/WilliamParker/6b927972aac07dd83ac98677e26ca589).

I believe that what is happening here is that NegationWithJoinFilterNode has a [get-elements](https://github.com/cerner/clara-rules/blob/0.15.0/src/main/clojure/clara/rules/memory.cljc#L440)  call [inside a for call which creates a lazy sequence](https://github.com/cerner/clara-rules/blob/0.15.0/src/main/clojure/clara/rules/engine.cljc#L829).  This get-elements call will retain the reference to a TransientLocalMemory even after [a to-persistent! call after fire-rules](https://github.com/cerner/clara-rules/blob/0.15.0/src/main/clojure/clara/rules/engine.cljc#L1949) is done [calls persistent! on the alpha-map field](https://github.com/cerner/clara-rules/blob/0.15.0/src/main/clojure/clara/rules/memory.cljc#L918).  It appears that we are not realizing this lazy sequence before passing it downstream.  Adding println calls in the restrictions suggests that we fully firing the rules without the sequence being realized.

Any fix should consider that while this particular case threw an exception, if we are not fully realizing the lazy sequences created in various places in the engine we are also vulnerable to mutation in the memory due to later processing steps being reflected in earlier processing steps; see https://github.com/cerner/clara-rules/issues/184.

**Comment by mrrodriguez** [comment]
This is a good fine. The obvious quick fix is to just try to eliminate the few offending lazy seqs and see if that causes any further issues.

However, considering a bit more sophisticated approach such as #305 definitely sounds like an interesting idea that could help avoid future issues, as well as possibly even help improve some performance scenarios.

**Comment by WilliamParker** [comment]
"The obvious quick fix is to just try to eliminate the few offending lazy seqs and see if that causes any further issues." 

I'm inclined to try just eliminating the laziness throughout the engine and see what the performance impacts are, and go with that if they're reasonable.  I suspect that the impact on performance will be minimal, since in most cases I'd expect realization to be forced eventually. For example, queuing activations should force realization of lazy sequences in Clojure because of the use of non-lazy Java queues, and even in ClojureScript the [use of into](https://github.com/cerner/clara-rules/blob/0.15.0/src/main/clojure/clara/rules/memory.cljc#L787) would cover many cases.  Note that even if realization is forced before we return from fire-rules, there is still a possibility of incorrect rules outcomes although the exception in this issue won't be thrown.

Overall, trying to combine laziness with the expectations of the memory around mutability is a complex and error-prone endeavor that I'd rather not deal with if we aren't gaining much from it.  The other thing I've been mulling on in relation to this is that if we do need some form of laziness, we might be better off exploring unlinking as described in various places in the Rete literature including Doorenbos's paper.  There may be other similar sorts of paths we could pursue as well.  Any laziness we're relying on for performance from "for" is somewhat accidental and deliberate domain-specific laziness like unlinking would probably be more robust.

@rbrush did you specifically write the for loops in the nodes in the engine with the expectation of using laziness?

**Comment by rbrush** [comment]
Sorry for the very delayed response...I had somehow overlooked this thread. The laziness here wasn't a purposeful decision, but rather just _me_ being lazy and using familiar for comprehensions in the logic. Making this logic eager makes sense to me.

I am surprised that the lazy sequence isn't actually being realized _before_ persistent! is called (in which case this wouldn't be a problem)...since typically I'd expect the following flow through the engine:

* The memory becomes transient
* We work with it until it stabilizes and no further changes are pending
* The memory becomes persistent

But I suppose there are internal structures to the memory that can become persistent in intermediate steps as we go, which is probably what's happing here.

**Comment by WilliamParker** [comment]
@rbrush I modified the test slightly to demonstrate what I think is happening:

````
(def-rules-test test-negation-complex-join-with-numerous-non-matching-facts-inserted-after-descendant-negation
  {:rules []
   :queries [query1 [[]
                     [[Hot (= ?t temperature)]
                      [:not [Cold (join-filter-equals ?t temperature)]]]]]
   :sessions [empty-session [query1] {}]}
  (let [lots-of-hot (doall (for [_ (range 100)]
                             (->Hot 20)))]
    (def s (-> empty-session
               (insert (->Cold 10))
               fire-rules
               (insert-all lots-of-hot)
               fire-rules))))
````

Obviously I realize this isn't much of a test since it isn't asserting anything, but it makes the session available in the REPL.  This test ran without errors. I first established that the QueryNode has a node-id of 3.

````
(-> s .rulebase :query-nodes)
````

I then accessed the sequence of tokens for the query.  Simply checking the type of the lazy seq didn't cause an error, but attempting to realize it did.

````
clara.test-memory> (-> s .memory :beta-memory (get 3) (get {}) type)
clojure.lang.LazySeq
clara.test-memory> (-> s .memory :beta-memory (get 3) (get {}))
({:matchesIllegalAccessError Transient used after persistent! call  clojure.lang.PersistentArrayMap$TransientArrayMap.ensureEditable (PersistentArrayMap.java:457)
clara.test-memory> 
````

Basically the sequence of tokens added [here](https://github.com/cerner/clara-rules/blob/0.15.2/src/main/clojure/clara/rules/engine.cljc#L434) isn't being fully realized, even after the session finishes firing the rules.  Since a LazySeq is a persistent Clojure collection it won't be modified by [->persistent-coll](https://github.com/cerner/clara-rules/blob/0.15.2/src/main/clojure/clara/rules/memory.cljc#L904) when making the memory persistent again.  Note that we don't start using mutable Java collections in the Clojure memory until we are combining multiple collections of the type in question, in this case tokens; see [add-tokens!](https://github.com/cerner/clara-rules/blob/0.15.2/src/main/clojure/clara/rules/memory.cljc#L589).

For rules I'd expect the firing process to force realization, but for queries we don't do anything with them until a user requests the results.

**Comment by mrrodriguez** [comment]
Should  [`->persistent-coll`](https://github.com/cerner/clara-rules/blob/0.15.2/src/main/clojure/clara/rules/memory.cljc#L904) be modified to ensure transients have `persistent!` called on them?

I'm not sure there is a convenient pred for this, but 
```clj
(instance? clojure.lang.ITransientCollection coll) 
```
would give you this info.

Does that solve the problem @ https://github.com/cerner/clara-rules/issues/303#issuecomment-320693705

@WilliamParker

**Comment by WilliamParker** [comment]
@mrrodriguez I think that would prevent the error from that test, yes.  That said, the error could also be prevented by moving the [get-elements calls in NegationWithJoinFilterNode](https://github.com/cerner/clara-rules/blob/0.15.2/src/main/clojure/clara/rules/engine.cljc#L829) out of the :when predicate, and thus force them to occur before passing the lazy sequence of tokens on.  The sequence still wouldn't be realized but there wouldn't need to be further interaction with the memory after passing it on.

However, I don't think that really addresses the full scope of the issue.  More broadly, I'd suggest that engine nodes should not emit values, either to the transport or memory, that may change in value after such emission.  When we build lazy sequences off of operations on mutable lists, we're doing exactly that.  If the values do end up changing between emission and their use, we're violating the assumptions of the Rete algorithm.  If we want this contract, we either have to fully realize values derived from mutable datasets returned from the memory before passing those values onward or only return persistent data structures from the memory.

If the performance impacts are reasonable my suggestion would be to simply replace all instance of for loops in the engine with (doall (for ....)).  In practice, I think we're probably fully realizing these sequences most of the time since

- Anything that feeds into a rule, as opposed to a query, will need to be fully realized by fire-rules.
- [Anything with join bindings will be forced to be realized by propagate-items-to-nodes](https://github.com/cerner/clara-rules/blob/0.15.2/src/main/clojure/clara/rules/engine.cljc#L153).
- [In Clojure, the seq call in the case with no join bindings in propagate-items-to-nodes](https://github.com/cerner/clara-rules/blob/0.15.2/src/main/clojure/clara/rules/engine.cljc#L163) will [realize the first 32 items](http://blog.fogus.me/2010/01/22/de-chunkifying-sequences-in-clojure/).
- On the JVM, once you transition to a mutable collection for a given node/binding group combination in the memory you'll obviously realize anything added into that collection.
- I'd expect that anything hit hard by truth maintenance operations would have to mostly realize things anyway due to the need to remove arbitrary elements from the middle of the sequence.  For context, these operations could in principle use hash-based collections but we found that in practice highly optimized linear sweeps through a list were faster.

If this is the case, then the performance impact will probably be minimal.  The cases that I can most easily see being harmed would be use-cases with complex filters in queries that don't experience many truth maintenance operations. 

To give an example of a potentially problematic workflow, suppose that a fact is propagated into the network through unrealized lazy sequences and then retracted before those sequences are realized.  It could be "removed" from downstream nodes by mutation at its entry into the rules network while subsequently still propagating through the network as a retraction.  This could pose problems, particularly if there are duplicates of the fact that were not retracted and should remain in the network.   Overall, it is difficult to reason about how these things will play out when every part of the engine can interact with any other part in this manner, and I'd prefer to avoid it if we can.

I've toyed with the idea of working to find or create a data structure that would give us some of the performance benefits of a mutable list in the memory while allowing us to take immutable "snapshots"; since we wouldn't need efficient immutable changes on top of those snapshots there might be things we could do that Clojure's collection implementations can't, although I doubt we'd be able to get a "free lunch" i.e. we'd still lose some of the performance benefits of straight mutation.  I don't think we should take on that complexity without a concrete reason to do so though.

**Comment by mrrodriguez** [comment]
> However, I don't think that really addresses the full scope of the issue. More broadly, I'd suggest that engine nodes should not emit values, either to the transport or memory, that may change in value after such emission. When we build lazy sequences off of operations on mutable lists, we're doing exactly that. If the values do end up changing between emission and their use, we're violating the assumptions of the Rete algorithm. If we want this contract, we either have to fully realize values derived from mutable datasets returned from the memory before passing those values onward or only return persistent data structures from the memory.

I agree with that. We really just shouldn't entirely avoid mixing lazy evaluation and mutable "containers". Things have to be fully realized before entering into any lazy context. This is just basic principles of using lazy evaluation really.

> If the performance impacts are reasonable my suggestion would be to simply replace all instance of for loops in the engine with (doall (for ....)). In practice, I think we're probably fully realizing these sequences most of the time since

If that works without big consequence, that is a straightforward solution and +1 to that.

> I've toyed with the idea of working to find or create a data structure that would give us some of the performance benefits of a mutable list in the memory while allowing us to take immutable "snapshots"

It's an interesting thought. It isn't immediately clear to me how much it would help, but I haven't thought it through as much as you. However, if we can just stick with the built-in and possibly even have the best perf via mutate-in-place, then it'd be good to stick with. Well tested data structures, etc.


-----

Lastly, I do still wonder if [`->persistent-coll`](https://github.com/cerner/clara-rules/blob/0.15.2/src/main/clojure/clara/rules/memory.cljc#L904) should be written to ensure that all Clojure transients are made persistent. It just makes more sense for it to do that job. It isn't really doing the job of making a coll persistent if it leaves transients. The cljs side I believe does always ensure this. I think it may just lead to perhaps somewhat unrelated other issues later if transients can leak through.

**Comment by WilliamParker** [comment]
Circling back to a comment from a while back:

"Lastly, I do still wonder if ->persistent-coll should be written to ensure that all Clojure transients are made persistent. "

I think we discussed this offline a while back, but this wouldn't solve the problem of having laziness with underlying mutable collections during rules firing.  In any case, aggressively realizing all relevant lazy sequences had not noticeable performance impacts.

This has been fixed by https://github.com/cerner/clara-rules/pull/335.  Are there any objections to closing this issue?

**Comment by mrrodriguez** [comment]
@WilliamParker 
I understand that the `->persistent-coll` fn wasn't the issue entirely. My comment was about the clarity of what that fn does and to prevent future issues.

I think the `->persistent-coll` impl would be improved by ensuring that it actually does return a persistent collection, not a transient & not mutable coll from Java, etc.

It isn't a huge issue, but could prevent other issues like this from creeping in later. I agree that the bigger issue was solved with this #335.

I have no issue with it being closed.

**Comment by WilliamParker** [comment]
I've linked to Mike's previous comment from https://github.com/cerner/clara-rules/issues/305; this sort of extra validation discussion belongs either there or an another issue IMO.

Closing.
